### ISSUE TITLE
Class cast exception with the agent

### ISSUE THREAD

**Issue by edinardo** [DataDog/dd-trace-java#3079]
We are seeing `java.lang.ClassCastException` with the message `'datadog.trace.bootstrap.instrumentation.java.concurrent.ComparableRunnable cannot be cast to java.util.concurrent.Delayed' @ java.util.concurrent.DelayQueue.add( )`, usually related to JDBC calls. Have you seen this before?

We are using Datadog agent version 0.87.0 without extra parameters besides service name and version with Spring Boot version 2.1.8.RELEASE.

Thanks in advance!

**Comment by mcculls** [comment]
Hi @edinardo can you share more of the stacktrace? Also which JDBC driver are you using?

**Comment by edinardo** [comment]
We are using jTDS version 1.2.8 and the stack trace:
```
java.sql.SQLException: java.lang.ClassCastException: datadog.trace.bootstrap.instrumentation.java.concurrent.ComparableRunnable cannot be cast to java.util.concurrent.Delayed
        at com.gameaccount.proc.execution.StoredProcedureExecutor.obtainConnection(StoredProcedureExecutor.java:116)
        at com.gameaccount.proc.execution.StoredProcedureExecutor.doExecute(StoredProcedureExecutor.java:74)
        at com.gameaccount.proc.execution.StoredProcedureExecutor.execute(StoredProcedureExecutor.java:45)
        at com.gameaccount.proc.execution.cache.CachingStoredProcedureExecutor.execute(CachingStoredProcedureExecutor.java:62)
        at com.gameaccount.proc.proxy.StoredProcedureInvocationHandler.doInvoke(StoredProcedureInvocationHandler.java:73)
        at com.gameaccount.proc.proxy.InvocationHandler.doInvoke(InvocationHandler.java:28)
        at com.gameaccount.commons.proxy.handler.InvocationHandler.invoke(InvocationHandler.java:27)
        at com.sun.proxy.$Proxy139.execute(Unknown Source)
        at com.gameaccount.leaderboards.type.repository.DbLeaderboardTypeRepository.loadLeaderboardTypes(DbLeaderboardTypeRepository.java:172)
        at com.gameaccount.leaderboards.type.repository.DbLeaderboardTypeRepository$$FastClassBySpringCGLIB$$f1df41b.invoke(<generated>)
        at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218)
        at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:749)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
        at org.springframework.dao.support.PersistenceExceptionTranslationInterceptor.invoke(PersistenceExceptionTranslationInterceptor.java:139)
        at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:186)
        at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688)
        at com.gameaccount.leaderboards.type.repository.DbLeaderboardTypeRepository$$EnhancerBySpringCGLIB$$565239af.loadLeaderboardTypes(<generated>)
        at com.gameaccount.leaderboards.type.service.LeaderboardTypeManagementServiceImpl.getAllTypes(LeaderboardTypeManagementServiceImpl.java:247)
        at com.gameaccount.leaderboards.type.controller.LeaderboardTypeManagementController.getTypes(LeaderboardTypeManagementController.java:65)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.springframework.web.method.support.InvocableHandlerMethod.doInvoke(InvocableHandlerMethod.java:190)
        at org.springframework.web.method.support.InvocableHandlerMethod.invokeForRequest(InvocableHandlerMethod.java:138)
        at org.springframework.web.servlet.mvc.method.annotation.ServletInvocableHandlerMethod.invokeAndHandle(ServletInvocableHandlerMethod.java:104)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.invokeHandlerMethod(RequestMappingHandlerAdapter.java:892)
        at org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerAdapter.handleInternal(RequestMappingHandlerAdapter.java:797)
        at org.springframework.web.servlet.mvc.method.AbstractHandlerMethodAdapter.handle(AbstractHandlerMethodAdapter.java:87)
        at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1039)
        at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:942)
        at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1005)
        at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:897)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:634)
        at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:882)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:741)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:231)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.AbstractRequestLoggingFilter.doFilterInternal(AbstractRequestLoggingFilter.java:262)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.boot.actuate.web.trace.servlet.HttpTraceFilter.doFilterInternal(HttpTraceFilter.java:88)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:99)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:92)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:93)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at datadog.trace.instrumentation.springweb.HandlerMappingResourceNameFilter.doFilterInternal(HandlerMappingResourceNameFilter.java:50)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.filterAndRecordMetrics(WebMvcMetricsFilter.java:114)
        at org.springframework.boot.actuate.metrics.web.servlet.WebMvcMetricsFilter.doFilterInternal(WebMvcMetricsFilter.java:104)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:200)
        at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:118)
        at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:193)
        at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:166)
        at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:202)
        at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:96)
        at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:526)
        at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:139)
        at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
        at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
        at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:343)
        at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:408)
        at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:66)
        at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:860)
        at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1587)
        at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:49)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassCastException: datadog.trace.bootstrap.instrumentation.java.concurrent.ComparableRunnable cannot be cast to java.util.concurrent.Delayed
        at java.util.concurrent.DelayQueue.add(DelayQueue.java:70)
        at com.gameaccount.datasource.internal.standard.task.scheduler.ExecutorTaskScheduler$1.execute(ExecutorTaskScheduler.java:88)
        at com.gameaccount.datasource.internal.standard.task.scheduler.ExecutorTaskScheduler.schedule(ExecutorTaskScheduler.java:37)
        at com.gameaccount.datasource.internal.standard.lease.manager.LeaseManager.scheduleForExpiry(LeaseManager.java:94)
        at com.gameaccount.datasource.internal.standard.lease.manager.LeaseManager.createLease(LeaseManager.java:43)
        at com.gameaccount.datasource.internal.connection.managed.ManagedConnection.lease(ManagedConnection.java:102)
        at com.gameaccount.datasource.ConnectionPool.acquire(ConnectionPool.java:241)
        at com.gameaccount.datasource.DataSourceImpl.getConnection(DataSourceImpl.java:27)
        at com.gameaccount.proc.execution.StoredProcedureExecutor.obtainConnection(StoredProcedureExecutor.java:108)
```

**Comment by bantonsson** [comment]
Hi @edinardo, and sorry for the late reply. We've been working on a fix for this, and there is currently a PR #3468
